### PR TITLE
jump function for Mersenne Twister

### PR DESCRIPTION
**Pull request by wildart**
This PR adds `mtjump` function for Mersenne Twister RNG which moves RNG state forward without generating numbers.

### CODE REVIEW

File: base/docs/helpdb.jl [hunk 1]
```diff
@@ -16340,3 +16340,14 @@ If ``pkg`` is installed, return the installed version number, otherwise return `
...
 Pkg.installed
 
+doc"""
+```rst
+::
+           mtjump(seed, jumps, [jumppoly]) -> Vector{MersenneTwister}
+
+Create an array of the size ``jumps`` of initialized ``MersenneTwister`` RNG objects where the first RNG object in the array is initialized with ``seed`` and following ``MersenneTwister`` RNGs in the array initialized such that a state of the RNG object in the array would be moved forward (without generating numbers) from a previous RNG object array element on a particular number of steps encoded by the jump polynomial ``jumppoly``.
+
+Default jump polinomial moves forward ``MersenneTwister`` RNG state on 10^20 steps.
```
Comment (IainNZ) [inline base/docs/helpdb.jl]: polynomial

**Comment by ViralBShah** [comment]
This also should have an entry in NEWS.

Cc: @rfourquet

**Comment by ViralBShah** [comment]
I guess the Appveyor build needs restarting.

**Comment by tkelman** [comment]
This seems a little too niche to be worth exporting. And added functions need to have signatures added for them somewhere in the RST docs for `doc/genstdlib.jl` to correctly populate docstrings.

File: base/dSFMT.jl [hunk 2]
```diff
@@ -5,11 +5,24 @@ module dSFMT
...
...f9755085ccc8456a3090b73a3031a201d9c704a4ad4868dd9b6d06205560013973f60d637de2f18354bf4523d9d81dc2a7e78cd42c586364bbe0ed86fde0f081f801c1a4abb830839b7796d9a01f141bec8bd93144104c6dc59170162c0a5a639eb63a0a164970de50eb2e04f027394b26ed48d341f7851994df79d7cd663672a556f25e5e16a3adbe1003d631de938fabfed234df12b5ff3027f4a2da823834cb098e0f977a4eb9614579d5c7a1d400a1a933a657aef8ea1a66743d73b0cf37a7d64e9a63e4c7b09945f0db750b311b39783fb5ea216616751967d480a630d3da7c89d1c7beae20369137e96734a4cfedca56a7887f076fe4fe97534ad3e4f74d1a81750581a5ea214b440c7f30331ab86c257534c71175d1e731303a48b01c589fda4fb0d4368b4dd63d91204cb6fc389b2202aa94391907bfb72902a4031f5589ed5f391c2ce92aa998c200ba3c77d8bd747b9d0a29fa85cda3949a6d2bd0c3402e68f98fd451aa27b6c2dfd170e004577cbdb25e3a1b9852e9f66a370789c47bfce722446dade1b32ceae71ee0e1d96edf7ed08a93e3690056f46c3d8e63f88e53673ee71d72cfedbeba493ee91333120e09e9ce9f9c9a7a400f814ea618b1de48f9805e092f4e20f301fbb65caa83735a2a5c89befe4bce4116dca3688e1e14c6f09a945671dedbb5c0ba526842b6cae31d8b5ff978bae928a17a75c134630dd9de988f6ad3d89a071b33775a9660a40b48ec61ad3f93ac81cb1c65d8b0bab5c214786abd13cc10a8ea2e2a370e86e2fa1a372d83c9697b5e37b281e51507685f714fdaebe49ffc93a5582e1936eaee8e4140a4b72"
```
Comment (tkelman) [inline base/dSFMT.jl]: how long is the code that generated this?

File: test/random.jl [hunk 3]
```diff
@@ -357,3 +357,24 @@ let mta = MersenneTwister(42), mtb = MersenneTwister(42)
...
...4f1cea849c12dc4a3d38d3010567f387590532b78927e92f0b718c84e882b3df071a78a011d0fd56d4101dcb009914a16a781b240a6fb2440c72b0ffb365be9d3459d114e665a0d35d7b8bd280101d85d1211d939ba0b15ab528c4f9dd2b001172561d211671b96873010ae3c2b8317f773d735698914228764b831423ae19dd4bbb008b9f1bd1e3ebdd626e629a46a9dd70bdd4bb30e2279e83c12bbbead6479b5f9980b1a9c785395520703a9367d931b45c1566c9d314b1745cafc6d0667cc9bc94d0c53a960c829eb09b768ab6bb2133e4fea1d939f3d3f8e1237210cf3577c830f0493073dc1d189abf27402b8b31b7c172c43dbf331a0828adfe737380e763d0ab0bfaaf94ec04830f94380a83718f340c4eeb20d7eb22b94613be84a9ed332ab364efff6cb37eec35d186185cca725e7a748f6bdb427604fb1628d49a7424a5a62a2e930fe142b035503af332fe748d5e63591b9ac54071ca843d5e474a48837de8b80387f3269ab50d2fd99c08c971e015d13fa02c7c315922ce58bdacbf8ee48827851a61fca59882d7eadcce3166dfe012aa9ec849e698e776a4d384f4755b506a222636942a81bbbffa1ff47e4d81fe68120aebcfd1a7e0000fd0cffdc44e1f0cd69ea2b4936564c78af51fed1cc8e34f0b46d6330b4b50ddee09335b7b0be0bc9f7f8e48415e15d08f811653d21bc6dd152742b086caadcc6dff5e27b40da42c2f1ebf3dd2bd51c418718e499859239317fcab10892eadf1c0ebf7a4246bce4cce3617193032f3e41b977dc8650298ac39631c527460364effea0f0bfd043df72ead0406aba1bcd636d65d7b89979eb8e1";
```
Comment (tkelman) [inline test/random.jl]: likewise here

**Comment by ViralBShah** [comment]
Yes, we probably should not export dsfmt_jump, and have it as an internal API. Parallel rand will start using it right away, and we can stabilize a user facing API in 0.5.

I guess one possibility is randjump(r::AbstractRNG, howmuch).

File: base/dSFMT.jl (hunk 2 above)
Comment (wildart) [inline base/dSFMT.jl]: There is a special tool that generates jump polynomials. All info [here](http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/SFMT/JUMP/dsfmt-jump.html).

**Comment by wildart** [comment]
> we probably should not export dsfmt_jump

Agree.

> I guess one possibility is randjump(r::AbstractRNG, howmuch).

Bare in mind this functionality works **only** with `MersenneTwister`, to be exact with dSFMT, that is why I called in `mtjump`.

@ViralBShah Parallel `rand` should be straight forward - change GLOBAL_RNG state for every worker. So we could `prand(seed, [jumppoly])` and number of jumps equals to number `nprocs`

**Comment by ViralBShah** [comment]
I did mean to ask - is jumping a MT special feature? `mtjump` that is not exported sounds like the right name. I was hoping to immediately bake in the parallel rand after this was merged! Even there, not sure what the API would be, but the capability is certainly that easy.

**Comment by wildart** [comment]
> is jumping a MT special feature?

Yes, as I understand, it is applicable only to MT, specifically to dSFMT implementation.

**Comment by toivoh** [comment]
Jumping should be feasible for any linear random number generator with a
reasonable size state space, by applying power by squaring to an n x n
matrix över GF(2), where n is the number of bits of state. I explored this
for the xorshift rng algorithm (n = 4096) some time ago, and in that case
it could be done quite fast.
Jumping is of course also trivial for random123.

So not having done any actual investigation, I would hazard a guess that
jumping can be implemented reasonably efficiently for most rngs.

**Comment by wildart** [comment]
@toivoh Does it mean that different RNG types would have different jumping algorithms, if any?

**Comment by toivoh** [comment]
Any linear one could reuse the same power by squaring code if you just
write a little bit of specific code that constructs the original iteration
matrix. (This applies to mt too, but it has a pretty large state space and
I'd guess that your code is quite a lot more efficient. Don't know if the
principles it is based on is applicable to other rngs as well.) Rngs that
are not linear would need custom jumping algorithms in general.

**Comment by ViralBShah** [comment]
It does seem then, not to have `mtjump`, and instead have `randjump`, where other jump algorithms could potentially be implemented later.

**Comment by wildart** [comment]
I renamed it to `randjump`.

File: base/docs/helpdb.jl [hunk 4]
```diff
@@ -16340,3 +16340,14 @@ If ``pkg`` is installed, return the installed version number, otherwise return `
...
+```rst
+::
+           randjump(r::MersenneTwister, jumps, [jumppoly]) -> Vector{MersenneTwister}
+
+Create an array of the size ``jumps`` of initialized ``MersenneTwister`` RNG objects where the first RNG object in the array is initialized with ``seed`` and following ``MersenneTwister`` RNGs in the array initialized such that a state of the RNG object in the array would be moved forward (without generating numbers) from a previous RNG object array element on a particular number of steps encoded by the jump polynomial ``jumppoly``.
+
+Default jump polynomial moves forward ``MersenneTwister`` RNG state on 10^20 steps.
+```
+"""
+randjump
```
Comment (tkelman) [inline base/docs/helpdb.jl]: I think you need to add a signature somewhere in the RST docs for `doc/genstdlib.jl` to populate the docstring correctly

**Comment by andreasnoack** [comment]
You could argue the the `rand` part is unnecessary in the function name since the first argument is an `AbstractRNG`.

The MT can potentially end up in bad states where the produced numbers are not good. The MT library we use tries to select a good seed vector from the user provided seed, but do you know what the literature says about jumping a long step forward? I'll try to run the RNG tests with a jumped seed vector a couple of times and see how it goes.

**Comment by wildart** [comment]
@andreasnoack Because MT has very-very long period (2^19937-1), it's quite save to do a long jumps.

**Comment by andreasnoack** [comment]
My reasoning was the following. Assume for simplicity that the development of the state in the MT can be expressed as `bbbgggbbbggg` where `g` is good and `b` is bad and because the period is long then a single letter represents the number of draws my machine can make in 1000 years or something like that. Now the seed function in the MT library will ensure that we end up in a `g` epoch and even if I draw quite a few numbers there, I won't make it out the good epoch. However, I made a back-on-the-envelope calculation and one of your jumps is equivalent to something like 8000 years of draws on my machine so I'm wondering if we can then end up in a `b` epoch.

**Comment by wildart** [comment]
@andreasnoack I didn't understand  about a good and bad part of the MT state. Period defines repetition-free sequence of numbers, and partially reflects RNG quality. It has nothing to do with seeding because RNG seed only initializes particular sequence of random numbers. So, when RNG initialized with some seed to produce particular sequence, jumping marks several subsequences in it, and produces initial states for multiple RNGs so they start running with each own subsequence. So, every RNG initialized from jumping procedure uses same seed, the only difference is number of steps went forward for every RNG. In the very long period there is a good chance to pick a several sequential subsequences that would never overlap. Non-overlapping is required for a parallel RNG initialization. Initializing multiple RNGs with different seeds   doesn't not guaranty that produces number sequences wouldn't overlap.

**Comment by jakebolewski** [comment]
Why can't parallel random number generation live in an external package?

**Comment by ViralBShah** [comment]
This is not just about parallel RNG. One can have multiple streams as well - with the jump capability. Anything can live in an external package, but this is a small amount of code that would be greatly useful for a number of purposes.

**Comment by andreasnoack** [comment]
[This](http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/MT2002/emt19937ar.html) is a discussion of the possibility of "bad" initialization. The standard example is when the state vector has many zeros, e.g.
```jl
julia> r = MersenneTwister();

julia> fill!(r.state.val, 0);

julia> r.state.val[1] = 1;

julia> rand(r, 20)
20-element Array{Float64,1}:
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
 -1.0
```
I saw a journal article arguing that a bad epoch could happen again after a while, but I don't know if that is a real concern. I'll try to run the test suite a couple of times on jumped MTs and if they only fail the usual test the MT fails then I think we are as fine as we can get.

**Comment by wildart** [comment]
@andreasnoack It isn't fair example. You just zeroed RNG state, of course it will produce bad sequence. We never manually set state of dSFMT. Why would you do that if dSFMT library seeding functions available?

**Comment by wildart** [comment]
I added documentation to RST docs. @ViralBShah Ready to merge.

**Comment by andreasnoack** [comment]
@wildart I just wanted to give an example of a state that produced bad random numbers. The intializer function ensures a good state, but is a good state + a jump (i.e. 8000 years of random numbers) also a good state? Maybe!

**Comment by wildart** [comment]
The answer would depend on RNG quality. For example, MT guarantees very long sequence of non-repeating numbers as if they were drawn from uniform distribution given that RNG state was properly set. Any states produced by jumping share same properties as the initial state. So if you trust your RNG and seeding procedure, you shouldn't have problems.

**Comment by andreasnoack** [comment]
> Any states produced by jumping share same properties as the initial state

This was basically what I have been questioning. Since it is known that the MT can have bad periods we'll eventually end one (maybe after 10k years or more) even when we start out from a good state. Normal use might be safe, but as mentioned above, a jump could be equivalent to 8000 years of random number generation. Is it documented in the dSFMT jump that we are safe here?

**Comment by jakebolewski** [comment]
@wildart, @ViralBShah how you initialize the jump sequence for parallel random number generation?  Looking at the code in [KMATH_RANDOM](http://www.aics.riken.jp/labs/lpnctrt/KMATH_RANDOM_e.html) library I'm wondering why initalizing the random state is so involved.

**Comment by andreasnoack** [comment]
@wildart I might sound more pessimistic here than I am. It is great that you have written this feature and hoprefully we'll soon have working parallel RNG. I'm just courious if the problem mentioned can happen with jumps. The testing I've promised has been delayed by local machine problems here at MIT, but hopefully I'll soon have server time to run the tests on.

**Comment by StefanKarpinski** [comment]
@andreasnoack, my very limited understanding is that once you've seeded MT correctly, jumping shouldn't be problematic since the "bad seed" problem seems to involve getting into a state where the normal period of 2^19937-1 doesn't apply.

**Comment by ViralBShah** [comment]
@jakebolewski See dsfmt jump here, where they have examples of jumping. Looks pretty straightforward in the case of dsfmt.

http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/SFMT/JUMP/dsfmt-jump.html

**Comment by ViralBShah** [comment]
I know that `randjump` is a bit redundant, but I like all the rand functions to have that prefix, and I don't yet feel like taking the name `jump` in base.

**Comment by andreasnoack** [comment]
After some bugs fixes and machine problems I got my RNG test package to run on some jumped rngs and their did no worse than MT without jumps so the jump concern might be purely theoretical.

**Comment by ViralBShah** [comment]
This is nice to know. Thanks for running the tests.

**Comment by rfourquet** [comment]
I am a bit late on this, but just wanted to say that I reviewed the code, and re-generated the polynomial using the code at http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/SFMT/JUMP/dsfmt-jump.html; and the streams of random numbers using the jump function in C or in Julia are similar in the few tests I did. So everything looks fine! At least as far as the original code in C is correct. This prompts my next question: the C code on which this PR is based off is "3-clause BSD License", are we fine with that? (otherwise we could then simply include this little C file in the dSFMT deps and create bindings to them). 

Also, would we want one or two more jump polynomials readily available? I have not yet a use for it, but could we imagine that a Random generator on a thread, made out of a jump, needs itself to create new generators by jumps (to get multi streams)? Just asking because it's some work to generate one (installing the software and understanding how to use it), but I can easily do so now.

File: base/dSFMT.jl (hunk 2 above)
Comment (rfourquet) [inline base/dSFMT.jl]: There seem to be an error in the name, shouldn't it be `JPOLY1e20` ?

File: base/docs/helpdb.jl [hunk 5]
```diff
@@ -16340,3 +16340,14 @@ If ``pkg`` is installed, return the installed version number, otherwise return `
 """
 Pkg.installed
 
+doc"""
+```rst
+::
+           randjump(r::MersenneTwister, jumps, [jumppoly]) -> Vector{MersenneTwister}
+
+Create an array of the size ``jumps`` of initialized ``MersenneTwister`` RNG objects where the first RNG object given as a parameter and following ``MersenneTwister`` RNGs in the array initialized such that a state of the RNG object in the array would be moved forward (without generating numbers) from a previous RNG object array element on a particular number of steps encoded by the jump polynomial ``jumppoly``.
```
Comment (rfourquet) [inline base/docs/helpdb.jl]: My brains has a hard time grasping this very very long sentence with no punctuation!

**Comment by ViralBShah** [comment]
It seems like a good idea to have better multi-stream support.

**Comment by wildart** [comment]
I think a small package that wraps jump polynomial creation would be more suitable rather than include this functionality into Base.

**Comment by ViralBShah** [comment]
The only reason for having this in base is to be able to have parallel RNGs work out of the box. Otherwise it could easily be a package.

**Comment by rfourquet** [comment]
OK, anyway I had prepared a PR yesterday, to see if there would be interest, so here it is!

**Comment by tkelman** [comment]
this was never added to NEWS for 0.4, was it https://github.com/JuliaLang/julia/pull/12498#issuecomment-128621769

**Comment by wildart** [comment]
We can add it to the NEWS when the update of RNG jump functionality, #16906, will be merged.

**Pull request merged**